File: answers/other/answers/lintcode/Search for a Range.cpp
class Solution {
    /** 
     *@param A : an integer sorted array
     *@param target :  an integer to be inserted
     *return : a list of length 2, [index1, index2]
     */
public:
    vector<int> searchRange(vector<int> &A, int target) {
        // write your code here
		int left(0);
		int right(A.size() - 1);
		while (left <= right && left >= 0 && right < A.size()) {
			int middle = (left + right) / 2;
			if (A[middle] > target) {
				right = middle - 1;
			}
			else if (A[middle] < target) {
				left = middle + 1;
			}
			else {
				while (A[left] < target && left < A.size()) {
					++ left;
				}
				while (A[right] > target && right >= 0) {
					-- right;
				}
				break;
			}
		}
		vector<int> range;
		if (left >= 0 && left < A.size() && A[left] == target) {
			range.push_back(left);
		}
		else {
			range.push_back(-1);
		}
		if (right >= 0 && right < A.size() && A[right] == target) {
			range.push_back(right);
		}
		else {
			range.push_back(-1);
		}
		return range;
    }
};

File: answers/leetcode/Search for a Range/Search for a Range.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] range = new int[2];
        range[0] = -1;
        range[1] = -1;
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (nums[middle] > target || (nums[middle] == target && middle > 0 && nums[middle - 1] == target)) {
                right--;
            } else if (nums[middle] < target) {
                left++;
            } else {
                range[0] = middle;
                break;
            }
        }
        
        left = 0;
        right = nums.length - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (nums[middle] > target) {
                right--;
            } else if (nums[middle] < target || (nums[middle] == target && middle + 1 < nums.length && nums[middle + 1] == target)) {
                left++;
            } else {
                range[1] = middle;
                break;
            }
        }
        return range;
    }
}